replicator/src/main/scala/com/evolutiongaming/kafka/journal/replicator/Batch.scala

```diff
@@ -14,139 +24,83 @@ private[journal] object Batch {
 
   def of(records: NonEmptyList[ActionRecord[Action]]): List[Batch] = {
 
-    // returns `true` when we can optimize Cassandra usage by NOT inserting "previous" Append actions in DB
-    // because "current" Delete action will discard them all
-    def cut(appends: Appends, delete: Action.Delete): Boolean = {
-      val append = appends.records.head.action
-      append.range.to <= delete.to.value
-    }
-
-    def dropDeleted(appends: Appends, delete: Action.Delete): Option[Appends] = {
-      // we cannot drop individual events in record as we do not parse the payload here (`records.head.action.payload: ByteVector`)
-      val validRecords = appends.records.collect { case record if record.action.range.to >= delete.to.value => record }
-      NonEmptyList.fromList(validRecords).map { Appends(appends.offset, _) }
-    }
-
-    records
-      .foldLeft(List.empty[Batch]) { (bs, record) =>
-        val offset = record.partitionOffset.offset
-
-        def appendsOf(records: NonEmptyList[ActionRecord[Action.Append]]): Appends = {
-          Appends(offset, records)
+    val state = records.reverse.foldLeft(State()) {
```

Comment:
Style comment: I would consider moving these `case .. =>` statements to `State` case class as methods. :blush:

Reply:
moved inside `State` - did it in two steps - at first created 3 extra functions (`handle*`), but then the code just asked for moving also the `switch` statement inside...